Implement more of Console on Unix

stephentoub opened the pull request:
This PR adds initial Unix implementations of:
- Console.Title
- Console.Beep
- Console.Clear
- Console.SetCursorPosition
- Console.WindowWidth / WindowHeight / WindowLeft / WindowTop
- Console.BufferWidth / BufferHeight
- Console.CursorLeft / CursorTop
- Console.KeyAvailable

As part of this, it also:
- Fixes how we set ICANON/ECHO modes, to leave the app in the new mode and only reset on exit
- Adds back several missing ConsoleKeys that had been commented out.
- Adds more ConsoleKeyInfo mappings so that more terminfo sequences are appropriately recognized
- Reduces allocations involved in sending ANSI escape strings
- Fixes a few minor issues, like places we were checking for null but should have been checking for null-or-empty

Note that I've not:
- Implemented the Windows versions of these.  That should be very straightforward, essentially just copying the code from the reference source.
- Added these to the System.Console contract.  We can do that once we a) agree that all of these should be added back, and b) provide the Windows implementations.

With these changes, I was able to compile and run several interesting console apps, such as a Space Invaders console game in C#, the LineEditor mentioned in https://github.com/dotnet/corefx/issues/4507, etc., as well as implement some of my own examples, like a progress bar.

cc: @weshaggard, @pallavit, @terrajobst, @ellismg 
Contributes to #4636

mellinoe commented on src/System.Console/src/System/ConsolePal.Unix.cs:
```diff
@@ -170,6 +253,129 @@ public static bool CursorVisible
+
+                int bytesRead = 0, i = 0;
+
+                // Response expected in the form "\ESC[row;colR".  However, user typing concurrently
```
On first glance, this seems like kind of scary code to me. Just to make sure I understand:

* User input can get accidentally consumed / ignored
* A user inputting digits can give you a weird/bad position back, at least for the individual call that it happens for.

How often have you seen any of the above actually happening? If the window for it to happen is something tiny like 0.1 milliseconds (or some other tiny interval), then there probably isn't that much to worry about, and the most likely bad thing to happen will be user input getting skipped.

stephentoub replied:
It is scary, but it's the nature of the beast: the protocol involves writing to stdout, at which point data gets put on stdin for reading.

Your understanding is correct, in that any other input that came in at the same time could be misintepreted.  We might also skip over such input.  I tried an approach where I put such data back into the ReadKey internal buffer so that it would subsequently be consumed, but then we potentially also end up with garbage data in the ReadKey buffer.

The window is small, but it's there.  It's most reproably if you hold down a key that generates an escape sequence, e.g. an arrow key, and have code that processes an arrow key by checking for the current position.  Then you end up having data showing up on stdin that's either from the escape sequence generated from pressing the arrow key or the escape sequence generated as a response to the current cursor position request.

I wouldn't expect much if any real input to get lost.  Any that did would be coming from someone typing, and it would likely be due to holding down a character key to have it repeating, in which case losing one or more isn't terrible as it's repeating.  Such concerns would also be mitigated by not checking for the current cursor position on every key request, which is what I was doing in my sample apps stressing this.  I have a TODO in the PR, but we could look into having some fields attempt to track the current cursor position, e.g. get the values once at the start of the app, and then based on the keys that are read, try to update those values to be kept in sync with the real cursor position rather than reading it every time.  That has its own risks, though.

mellinoe replied:
> the protocol involves writing to stdout, at which point data gets put on stdin for reading.

Yeah, it sounds like given the mechanism we have, this is the best we can do, other than tweaking things here and there.

> we could look into having some fields attempt to track the current cursor position, e.g. get the values once at the start of the app, and then based on the keys that are read, try to update those values to be kept in sync with the real cursor position

I was going to suggest something like this, as it seemed sort of similar to what we are now doing for ConsoleColor. On the other hand, it seems way riskier to do it for tracking the position because any tracking errors will accumulate, whereas the color gets "reset" every time you change it, so if you were out-of-sync with the previous color, it doesn't really matter now that you know the new color.

mellinoe commented:
> the protocol involves writing to stdout, at which point data gets put on stdin for reading.

Yeah, it sounds like given the mechanism we have, this is the best we can do, other than tweaking things here and there.

> we could look into having some fields attempt to track the current cursor position, e.g. get the values once at the start of the app, and then based on the keys that are read, try to update those values to be kept in sync with the real cursor position

I was going to suggest something like this, as it seemed sort of similar to what we are now doing for ConsoleColor. On the other hand, it seems way riskier to do it for tracking the position because any tracking errors will accumulate, whereas the color gets "reset" every time you change it, so if you were out-of-sync with the previous color, it doesn't really matter now that you know the new color.

khellang commented:
Yay! Soon we can run https://gist.github.com/XeeX/6220067 and https://gist.github.com/khellang/989aa3a947329661eab3 with corefx :grin:

stephentoub commented:
Thanks for the feedback, @ellismg.  I've addressed all of your comments.  I'm working on some additional improvements, but I'll do that as a separate PR.  Please let me know if you have any more thoughts or concerns on this PR.

ellismg commented on src/System.Console/src/System/IO/StdInStreamReader.cs:
```diff
@@ -69,34 +94,52 @@ public override string ReadLine()
+                else if (keyInfo.Key == ConsoleKey.Tab)
+                {
+                    _readLineSB.Append(keyInfo.KeyChar);
+                    Console.Write(' ');
```
Why do you write a Space here and not a `\t`?

stephentoub replied:
Because after writing a \t, if you try to backspace I don't then know how far back to move the cursor. I can investigate another solution separately.

ellismg replied:
Okay, makes sense.

ellismg commented:
LGTM

stephentoub commented:
Thanks for the review, Matt.

niemyjski commented:
Can this get reopened. I've been trying to get our sample console to work on osx as it does on windows and I've been having a heck of a time.

https://github.com/exceptionless/Exceptionless.Net/tree/master/src/Samples/Exceptionless.SampleConsole

SetCursorPosition doesn't seem to work at all and among other things the experience isn't even the same.  

On a side note, I can't even get our app to display anything in VS Code (output stream is directed to I think /dev/null 🎱 )...

niemyjski commented:
You can try a very basic .net core console app on osx and it won't work for SetCursorPosition...

danmosemsft commented:
@niemyjski It's probably better to open a new issue and reference this one.
